### PR TITLE
Add GLEAN

### PR DESCRIPTION
**Pull request by ckkelvinchan**
This PR adds `GLEAN` for image super-resolution. It makes use of `StyleGANv2` borrowed from MMGeneration. 

Paper: `GLEAN: Generative Latent Bank for Large-Factor Image Super-Resolution`

### CODE REVIEW

**Comment by ckkelvinchan** [comment]
Hi @Yshuo-Li, many of the changes are added by @nbei. Here are the main codes of GLEAN
1. `mmedit/models/backbones/sr_backbones/glean_styleganv2.py`
2. `mmedit/models/restorers/glean.py`

File: mmedit/models/backbones/sr_backbones/glean_styleganv2.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        self.style_channels = style_channels
+        channels = self.generator.channels
+
+        # encoder
+        num_styles = int(math.log2(out_size)) * 2 - 2
+        encoder_res = [2**i for i in range(int(math.log2(in_size)), 1, -1)]
+        self.encoder = nn.ModuleList()
+        self.encoder.append(
+            nn.Sequential(
+                RRDBFeatureExtractor(3, 64, num_blocks=23),
```
Comment (Yshuo-Li) [inline mmedit/models/backbones/sr_backbones/glean_styleganv2.py:127]: Maybe adding `mid_channels`(defaults 64) , `in_channels` and `out_channels` to the class (GLEANStyleGANv2) will be better.

File: mmedit/models/backbones/sr_backbones/glean_styleganv2.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+                    PixelShufflePack(
+                        in_channels, out_channels, 2, upsample_kernel=3))
+            else:
+                self.decoder.append(
+                    nn.Sequential(
+                        nn.Conv2d(in_channels, 64, 3, 1, 1),
+                        nn.LeakyReLU(negative_slope=0.2, inplace=True),
+                        nn.Conv2d(64, 3, 3, 1, 1)))
+
+    def forward(self, lr):
```
Comment (Yshuo-Li) [inline mmedit/models/backbones/sr_backbones/glean_styleganv2.py:181]: Named `LQ` in other models of MMEditing

File: mmedit/models/components/stylegan2/common.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        raise ValueError('The input module should contain parameters.')
+
+    if next(module.parameters()).is_cuda:
+        return next(module.parameters()).get_device()
+    else:
+        return torch.device('cpu')
+
+
+@torch.no_grad()
+def get_mean_latent(generator, num_samples=4096, bs_per_repeat=1024):
```
Comment (Yshuo-Li) [inline mmedit/models/components/stylegan2/common.py:25]: Need docstring

File: tests/test_glean.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                restorer(
+                    **data_batch,
+                    test_mode=True,
+                    save_image=True,
+                    meta=meta,
+                    iteration='1')
+        with pytest.raises(AssertionError):  # test with metric but gt is None
+            with torch.no_grad():
+                data_batch.pop('gt')
+                restorer(**data_batch, test_mode=True)
```
Comment (Yshuo-Li) [inline tests/test_glean.py:69]: Need unittests of the assertion errors

Comment (ckkelvinchan) [reply tests/test_glean.py:69]: The model for GLEAN should have no assertion errors. Do you mean the backbone? The unittests are in `test_glean_net.py`.

File: mmedit/models/components/stylegan2/common.py (hunk 3 above)
Comment (ckkelvinchan) [reply mmedit/models/components/stylegan2/common.py:25]: @nbei Could you help with that? I have copied the docstring of `get_mean_latent` from mmgeneration, but there is no docstring for `style_mixing` in mmgeneration.

File: tests/test_glean.py (hunk 4 above)
Comment (Yshuo-Li) [reply tests/test_glean.py:69]: Such as
mmedit/models/components/stylegan2/common.py#L16
mmedit/models/backbones/sr_backbones/glean_styleganv2.py#L271

Comment (ckkelvinchan) [reply tests/test_glean.py:69]: For the unittests for `glean_styleganv2.py`, it is in `test_glean_net.py`. I will add unittests for `common.py`.

**Comment by ckkelvinchan** [comment]
Will add docstring for `style_mixing`.

**Review by Yshuo-Li — APPROVED** [review]

File: mmedit/models/backbones/sr_backbones/glean_styleganv2.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+                pretrained=pretrained,
+                bgr2rgb=bgr2rgb))
+        self.generator.requires_grad_(False)
+
+        self.in_size = in_size
+        self.style_channels = style_channels
+        channels = self.generator.channels
+
+        # encoder
+        num_styles = int(math.log2(out_size)) * 2 - 2
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/glean_styleganv2.py:136]: For consistency, we use np for math ops

math.log2 -> np.log2

**Comment by innerlee** [comment]
Some notes:

* Replace all `math.xx` to `np.xx`
* For the style-gan code, we will refactor it to use mmgen once the infrastructure is ready. Currently we simply do the copy-paste

**Pull request merged**